File: fast_cp.py
import os
import random
import shutil
import sys
from multiprocessing import Pool

def copy_file(file, dest_dir):
    shutil.copy(file, dest_dir)

def main():
    src_dir = sys.argv[1]
    dest_dir = sys.argv[2]
    num_files = int(sys.argv[3])
    procs = int(sys.argv[4])

    files = [os.path.join(dp, f) for dp, dn, filenames in os.walk(src_dir) for f in filenames]
    random.shuffle(files)
    selected_files = files[:num_files]

    with Pool(processes=procs) as pool:
        pool.starmap(copy_file, [(file, dest_dir) for file in selected_files])

if __name__ == '__main__':
    main()

File: fast_cp.go
package main

import (
	"bufio"
	"math/rand"
	"os"
	"path/filepath"
	"sync"
	"time"
)

func main() {
	// Set the source directory
	srcDir := "ILSVRC2012/train/"

	// Set the destination directory
	destDir := "combined-msl-250k-imagenet-250k/ImageNet/"

	// Set the number of files to copy
	numFiles := 250000

	// Create a slice to store the file names
	files := make([]string, 0)

	// Walk the source directory and get a list of all the files
	filepath.Walk(srcDir, func(path string, info os.FileInfo, err error) error {
		if !info.IsDir() {
			files = append(files, path)
		}
		return nil
	})

	// Seed the random number generator
	rand.Seed(time.Now().UnixNano())

	// Use a wait group to keep track of the number of copying routines
	var wg sync.WaitGroup

	// Copy the specified number of random files
	for i := 0; i < numFiles; i++ {
		wg.Add(1)

		go func() {
			// Get a random file
			randIndex := rand.Intn(len(files))
			randFile := files[randIndex]

			// Create the destination file path
			destPath := filepath.Join(destDir, filepath.Base(randFile))

			// Copy the file using buffered I/O
			src, err := os.Open(randFile)
			if err != nil {
				wg.Done()
				return
			}
			defer src.Close()

			dest, err := os.Create(destPath)
			if err != nil {
				wg.Done()
				return
			}
			defer dest.Close()

			bufSrc := bufio.

